modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/planner/DataTypesPlannerTest.java

```diff
@@ -89,35 +90,47 @@ private void doTestSetOpNumbersCast(
                 for (String op : Arrays.asList("UNION", "INTERSECT", "EXCEPT")) {
                     String sql = "SELECT * FROM table1 " + op + " SELECT * FROM table2";
 
+                    if (log.isInfoEnabled())
+                        log.info("Test query: '" + sql + "', type1: " + t1 + ", t2: " + t2);
+
                     if (t1 == t2 && (!nullable1 || !nullable2))
                         assertPlan(sql, schema, nodeOrAnyChild(isInstanceOf(IgniteProject.class)).negate());
                     else {
                         RelDataType targetT = f.leastRestrictive(Arrays.asList(f.createSqlType(t1), f.createSqlType(t2)));
 
+                        // TODO https://issues.apache.org/jira/browse/CALCITE-7062 : Revert checks of NOT NULL casts after a fix.
+//                        assertPlan(sql, schema, nodeOrAnyChild(isInstanceOf(SetOp.class)
+//                            .and(t1 == targetT.getSqlTypeName() ? input(0, nodeOrAnyChild(isInstanceOf(IgniteProject.class)).negate())
+//                                : input(0, checkProject("TABLE1", "CAST($t0):" + targetT + (notNull ? " NOT NULL" : ""), "$t1")))
+//                            .and(t2 == targetT.getSqlTypeName() ? input(1, nodeOrAnyChild(isInstanceOf(IgniteProject.class)).negate())
+//                                : input(1, checkProject("TABLE2", "CAST($t0):" + targetT + (notNull ? " NOT NULL" : ""), "$t1")))
+//                        ));
+                        String cast1 = "[CAST($t0):" + targetT + ", $t1]";
+                        String cast2 = "[CAST($t0):" + targetT + " NOT NULL, $t1]";
+
                         assertPlan(sql, schema, nodeOrAnyChild(isInstanceOf(SetOp.class)
                             .and(t1 == targetT.getSqlTypeName() ? input(0, nodeOrAnyChild(isInstanceOf(IgniteProject.class)).negate())
```

Comment:
If now project merged into table, than IgniteProject check is useless, you need to check table with project.